src/modules/adaptors/Balancer/BalancerPoolAdaptor.sol

```diff
@@ -0,0 +1,295 @@
+     * TODO: rename params and possibly include others for `exit()` function. ex.) bptOut needs to be renamed to be agnostic.
+     * TODO: see issue for strategist callData
+     */
+    function useRelayer(
+        ERC20[] memory tokensIn,
+        uint256[] memory amountsIn,
+        ERC20 bptOut,
+        bytes[] memory callData
+    ) public {
+        for (uint256 i; i < tokensIn.length; ++i) tokensIn[i].approve(relayer(), amountsIn[i]); // TODO: this may be obsolete based on approval steps necessary for relayer functionality (see TODO below)
+
+        uint256 startingBpt = BPTout.balanceOf(address(this));
+
+        // TODO: implement proper way of giving Relayer approval; Relayer.setRelayerApproval() or Relayer.approveVault().
+
+        relayer().multicall(callData);
+
+        uint256 endingBpt = BPTout.balanceOf(address(this));
+
+        uint256 amountBptOut = endingBpt - startingBpt;
+
+        uint256 amountBptIn = priceRouter.getValues(tokensIn, amountsIn, bptOut); // TODO: not 100% sure how this math all works out. I think this roughly calcs the BPTValue In, it depends on the source of its priceData cause it has to know how much slippage to account for.
+
+        // check value in vs value out
+        if (amountBptOut < amountBptIn.mulDivDown(0.9e4, 1e4)) revert("Slippage");
+
+        // revoke token in approval
+        for (uint256 i; i < tokensIn.length; ++i) _revokeExternalApproval(tokensIn[i], relayer());
+
+        // TODO: see if special revocation is required or necessary with bespoke Relayer approval sequences
+    }
+
+    /**
+     * @notice deposit (stake) BPT into respective pool gauge
+     * @param _poolGauge address of rewards gauge
+     * @param _amountIn number of bpts to deposit
+     * @param _claim_rewards whether or not to claim pending rewards too (true == claim)
+     * @dev 1.) Interface custom as Balancer/Curve do not provide for liquidityGauges.
+     * @dev 2.) Strategists to provide _poolGauge (which they can get via off-chain querying methods)
+     * TODO: Finalize interface details when beginning to do unit testing
+     */
+    function depositBPT(address _poolGauge, uint256 _amountIn, bool _claim_rewards) external {
+        ILiquidityGaugev3Custom poolGauge = ILiquidityGaugev3Custom(_poolGauge);
+        IERC20 bpt = IERC20(poolGauge.lp_token());
+
+        if (bpt.balanceOf(address(this)) <= _amountIn) {
+            revert BalancerPoolAdaptor__NotEnoughToDeposit();
+        }
+        uint256 amountStakedBefore = poolGauge.balanceOf(address(this));
+
+        uint256 amountIn = _maxAvailable(address(bpt), _amountIn);
+        bpt.approve(_poolGauge, amountIn);
+        poolGauge.deposit(amountIn, address(this), _claim_rewards); // address(this) is cellar address bc delegateCall
+        uint256 actualAmountStaked = poolGauge.balanceOf(address(this)) - amountStakedBefore;
+        if (actualAmountStaked >= amountIn) {
```

Comment:
I dont think we need this check in here

Reply:
Yeah I think I did this in the wrong way anyways. We'd be more interested if the `amountStaked` was less than the amount we put in right? I'll remove the error though I wonder if we should still check that the `actualAmountStaked` was at least the amount we put in or a certain amount of it.

Reply:
Is pool guage upgradeable? If so then we can keep this check, otherwise the balanceOf logic is pretty simple so this check is not needed